internal/engine/tracer/context.go

```diff
@@ -143,9 +143,13 @@ func (c *context) StartOutput(ruleName string) Context {
 }
 
 func (c *context) start(component *enginev1.Trace_Component) Context {
+	components := make([]*enginev1.Trace_Component, len(c.components)+1)
+	copy(components, c.components)
+	components[len(c.components)] = component
```

Comment:
Would a `c.components = append(c.components, component)` not be (potentially) more efficient, rather than a full copy?

Reply:
Unless I am missing something, that's what we originally had? The problem is that `append` could modify the first argument in place if it has capacity so it's not guaranteed to return a new slice pointer every time.

Reply:
I think it's to do with the explicit return-reassign. It'll try to append contiguously if there's available memory, otherwise will reallocate behind the scenes ([docs](https://pkg.go.dev/builtin#append)):

```go
// The append built-in function appends elements to the end of a slice. If
// it has sufficient capacity, the destination is resliced to accommodate the
// new elements. If it does not, a new underlying array will be allocated.
// Append returns the updated slice. It is therefore necessary to store the
// result of append, often in the variable holding the slice itself:
//
//	slice = append(slice, elem1, elem2)
//	slice = append(slice, anotherSlice...)
//
// As a special case, it is legal to append a string to a byte slice, like this:
//
//	slice = append([]byte("hello "), "world"...)
func append(slice []Type, elems ...Type) []Type
```

Reply:
Got this wrong, it totally is re-assigning in the struct. This is about re-use of existing slices causing unpredictable results. From Charith:

>Yeah, it's rewriting the parent. So, say the path to the any expression is `a.b`. It then needs to go to child expressions `p`, `q` and `r` and for each them the parent component should be `a.b`. If we directly append to it from the children, then the result is unpredictable.